pkg/csr/csr.go

```diff
@@ -0,0 +1,260 @@
+	if len(c.Spec.Groups) < 2 {
+		return false, errors.Errorf("CSR %s contains invalid number of groups: %d", c.Name, len(c.Spec.Groups))
+	}
+	groups := sets.NewString(c.Spec.Groups...)
+	if !groups.HasAll(nodeGroup, systemPrefix) {
+		return false, errors.Errorf("CSR %s does not contain required groups", c.Name)
+	}
+
+	// Check usages include: digital signature, key encipherment and server auth
+	if !c.hasUsages(kubeletServerUsages) {
+		return false, errors.Errorf("CSR %s does not contain required usages", c.Name)
+	}
+
+	var hasOrg bool
+	for i := range csr.Subject.Organization {
+		if csr.Subject.Organization[i] == nodeGroup {
+			hasOrg = true
+			break
+		}
+	}
+	if !hasOrg {
+		return false, errors.Errorf("CSR %s does not contain required subject organization", c.Name)
+	}
+	return true, nil
+}
+
+// isNodeClientCert returns true if the CSR is from a  kube-apiserver-client-kubelet signer
+// reference: https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/#kubernetes-signers
+func (c *Approver) isNodeClientCert(x509cr *x509.CertificateRequest) bool {
+	if !reflect.DeepEqual([]string{nodeGroup}, x509cr.Subject.Organization) {
+		return false
+	}
+	if (len(x509cr.DNSNames) > 0) || (len(x509cr.EmailAddresses) > 0) || (len(x509cr.IPAddresses) > 0) {
+		return false
+	}
+	// Check usages include: digital signature, key encipherment and client auth
+	if !c.hasUsages(kubeletClientUsages) {
+		return false
+	}
+	return true
+}
+
+// hasUsages verifies if the required usages exist in the CSR spec
+func (c *Approver) hasUsages(usages []certificates.KeyUsage) bool {
+	if len(usages) != len(c.Spec.Usages) {
+		return false
+	}
+
+	usageMap := map[certificates.KeyUsage]struct{}{}
+	for _, u := range usages {
+		usageMap[u] = struct{}{}
+	}
+
+	for _, u := range c.Spec.Usages {
+		if _, ok := usageMap[u]; !ok {
+			return false
+		}
+	}
+
+	return true
+}
+
+// ParseCSR extracts the CSR from the API object and decodes it.
+func ParseCSR(csr []byte) (*x509.CertificateRequest, error) {
+	if len(csr) == 0 {
+		return nil, fmt.Errorf("CSR request spec should not be empty")
+	}
+	// extract PEM from request object
+	block, _ := pem.Decode(csr)
+	if block == nil || block.Type != "CERTIFICATE REQUEST" {
+		return nil, fmt.Errorf("PEM block type must be CERTIFICATE REQUEST")
+	}
+	return x509.ParseCertificateRequest(block.Bytes)
+}
+
+// areAddressesEqual returns true if the DNS address in CSR subject matches with the instance address
+// If the address passed is an IP address, we do a reverse lookup to find the host address
+func (c *Approver) areAddressesEqual(instanceAdd string, nodeName string) bool {
+	// reverse lookup the host if the address is an IP address
+	if parseAddr := net.ParseIP(instanceAdd); parseAddr != nil {
+		dnsAddresses, err := net.LookupAddr(instanceAdd)
+		if err != nil {
+			c.log.Info("failed to lookup hosts using", "IP", instanceAdd)
```

Comment:
Not quite sure about logging a "failed" with `Info` level.

Following the calls stack, [c.validateCSRContents()](https://github.com/openshift/windows-machine-config-operator/pull/499/files#diff-666ed6f6b4dbed560da68743f3814f61952cd552103ea310e478cde0c5cbed2eR76) throws an error if validated==false.

In that sense, returning an error here may be helpful, while reduce the burden of maintaining a boolean flag to control the execution flow.

Reply:
We do not want to return an error in cases where validation fails due to the instance address mismatch since the CSR could be from a linux instance.
 If validated == false we are only logging the error, the return errors.wrap [here](https://github.com/openshift/windows-machine-config-operator/pull/499/files#diff-666ed6f6b4dbed560da68743f3814f61952cd552103ea310e478cde0c5cbed2eR78)  would return nil if the err = nil.
I am trying to only return errors in CSR validation failure except the address mismatch.

Reply:
The error should be `failed to lookup DNS for IP`. 

What happens if we fail to reverse lookup an `instanceAddress` and it is not of a Linux node but that of a valid Windows instance whose CSR we should have approved? How are we going to convey that to the user?